### PR TITLE
feat: ledger signals and admin unenrollment buttons

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
**Description:**
Reasons behind the change: https://2u-internal.atlassian.net/browse/ENT-6935

TLDR - we need a way for support admins to revoke platform enrollment objects and choose whether or not to reverse the transaction object within the subsidy/ledger service. 

In implementing this we found the need to allow consumers to hook into certain events that can occur within the ledger models and services. As such I've added signals for un-enrollment related admin template/views and transaction reversal actions. In this way, the subsidy service can implement a receiver and use it's enterprise api client to update the appropriate LMS records upon the triggering of the appropriate admin unenroll event.      

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

File: openedx_ledger/views.py [hunk 1]
```diff
@@ -2,5 +2,96 @@
...
+        )
+
+    def post(self, request, transaction_id):
+        """
+        Handle POST request - handle form submissions.
+
+        Arguments:
+            request (django.http.request.HttpRequest): Request instance
+        """
+        print("sending admin invoked transaction unenroll signal")
```
Comment (adamstankiewicz) [inline openedx_ledger/views.py:46]: nit: should the print statement be removed?

File: openedx_ledger/views.py [hunk 2]
```diff
@@ -2,5 +2,96 @@
...
+            request,
+            self.template,
+            {'transaction': Transaction.objects.get(uuid=transaction_id)}
+        )
+
+    def post(self, request, transaction_id):
+        """
+        Handle POST request - handle form submissions.
+        """
+        print("Reversing transaction and sending admin invoked transaction unenroll signal")
```
Comment (adamstankiewicz) [inline openedx_ledger/views.py:83]: nit: Should this print statement be removed?

File: openedx_ledger/views.py [hunk 3]
```diff
@@ -2,5 +2,96 @@
...
+        transaction = Transaction.objects.get(uuid=transaction_id)
+        # TODO: determine how to generate idempotency key
```
Comment (adamstankiewicz) [inline openedx_ledger/views.py:85]: [curious] Is this still a relevant TODO for this PR or is it getting tackled elsewhere?

File: openedx_ledger/admin.py [hunk 4]
```diff
@@ -90,6 +93,52 @@ class Meta:
...
+        # url names coming from get_urls are prefixed with 'admin' namespace
+        manage_learners_url = reverse("admin:unenroll_with_refund", args=(obj.uuid,))
+        return HttpResponseRedirect(manage_learners_url)
+
+    unenroll_with_refund.label = "Unenroll & Refund"
+    unenroll_with_refund.short_description = (
+        "Reverse a transaction and unenroll the learner from the platform representation of the course."
+    )
+
+    def unenroll_without_refund(self, request, obj):
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply openedx_ledger/admin.py:111]: ahh well I bounced this back and forth in my head for a while, I'm fine with just unenroll

File: openedx_ledger/signals/signals.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+"""
+openedx_ledger related signals.
+"""
+
+from django.dispatch import Signal
+
+# Signal that indicates that a transaction object has been reversed.
+# providing_args=[
+#         'reversal',  # Reversal object
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply openedx_ledger/signals/signals.py:9]: https://github.com/openedx/openedx-ledger/blob/main/openedx_ledger/models.py#L467-L475 reversals are directly linked with their associated transactions, it feels a little redundant to add the transaction is as well

[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply openedx_ledger/signals/signals.py:9]: the admin invoked signal doesn't necessary always have a reversal associated with it as there's an option to unenroll without a refund

File: openedx_ledger/views.py (hunk 1 above)
Comment (alex-sheehan-edx) [reply openedx_ledger/views.py:46]: it certainly should, I'll swap it over to a log

File: openedx_ledger/views.py (hunk 3 above)
Comment (alex-sheehan-edx) [reply openedx_ledger/views.py:85]: That's there to remind me to discuss this with @johnnagro before this gets merged 😄

File: openedx_ledger/signals/signals.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply openedx_ledger/signals/signals.py:9]: Got it, thanks for the clarification!

File: openedx_ledger/views.py [hunk 6]
```diff
@@ -1,6 +1,103 @@
...
+        Returns:
+            django.http.response.HttpResponse: HttpResponse
+        """
+        return render(
+            request,
+            self.template,
+            {'transaction': Transaction.objects.get(uuid=transaction_id)}
+        )
+
+    def post(self, request, transaction_id):
```
Comment (pwnage101) [inline openedx_ledger/views.py:82]: Bunch of business logic below that should probably be in a shared utility function that can be re-used later for the `/api/v1/transactions/<>/reverse` REST API.

File: docs/decisions/0005-ledger-event-signals.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+`ADMIN_INVOKED_TRANSACTION_UNENROLL` is a unique signal that will be specifically sent when an admin user
+clicks the `unenroll` or `unenroll & refund` buttons within the transaction admin screen. This signal represents an
+override and cancelation of a transaction that requires the revokation of platform enrollment objects. This signal is
+unrelated to the state of a transaction record or it's potential reversal.
+
+Consequences
+************
+
+* Consumers of the ledger service will be able to hook into the reversal process and perform any actions they need to.
+* Patterns for signal events will be established and documented in the ledger service.
```
Comment (iloveagent57) [inline docs/decisions/0005-ledger-event-signals.rst:46]: nit: s/ledger service/subsidy service/

File: docs/decisions/0005-ledger-event-signals.rst [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+the `signals` dir of the openedx-ledger app. Starting out, the signals will be defined and named as follows:
+
+* `TRANSACTION_REVERSED`
+* `ADMIN_INVOKED_TRANSACTION_UNENROLL`
+
+although this list is not exhaustive and may change as the need arises.
+
+`ADMIN_INVOKED_TRANSACTION_UNENROLL` is a unique signal that will be specifically sent when an admin user
+clicks the `unenroll` or `unenroll & refund` buttons within the transaction admin screen. This signal represents an
+override and cancelation of a transaction that requires the revokation of platform enrollment objects. This signal is
```
Comment (iloveagent57) [inline docs/decisions/0005-ledger-event-signals.rst:39]: nit: s/revokation/revocation/

File: openedx_ledger/admin.py [hunk 9]
```diff
@@ -90,6 +93,52 @@ class Meta:
...
+        return HttpResponseRedirect(manage_learners_url)
+
+    unenroll.label = "Unenroll"
+    unenroll.short_description = (
+        "Unenroll the learner from the platform representation of the course without reversing the transaction object."
+    )
+
+    def get_urls(self):
+        """
+        Returns the additional urls used by the custom object tools.
```
Comment (iloveagent57) [inline openedx_ledger/admin.py:126]: question: what are "custom object tools"?

File: openedx_ledger/admin.py [hunk 10]
```diff
@@ -90,6 +93,52 @@ class Meta:
         readonly_fields = _all_fields
     inlines = [ExternalTransactionReferenceInlineAdmin]
 
+    change_actions = ('unenroll_with_refund', 'unenroll',)
+
+    def unenroll_with_refund(self, request, obj):
+        """
+        Redirect to the unenroll view.
```
Comment (iloveagent57) [inline openedx_ledger/admin.py:100]: nit: copypasta from docstring below?

Comment (alex-sheehan-edx) [reply openedx_ledger/admin.py:100]: whoops!

File: openedx_ledger/admin.py (hunk 9 above)
Comment (alex-sheehan-edx) [reply openedx_ledger/admin.py:126]: custom objects refers to `DjangoObjectActions` in this case

Comment (alex-sheehan-edx) [reply openedx_ledger/admin.py:126]: I can update the wording to be more clear

File: openedx_ledger/views.py [hunk 11]
```diff
@@ -1,6 +1,62 @@
...
+            "Reversing transaction and sending admin invoked transaction unenroll signal for transaction: "
+            f"{transaction_id}"
+        )
+        transaction = Transaction.objects.get(uuid=transaction_id)
+        # TODO: determine how to generate idempotency key
+        reversal = reverse_full_transaction(
+            transaction,
+            idempotency_key=f"admin-invoked-reverse-{str(transaction.uuid)}"
+        )
+        ADMIN_INVOKED_TRANSACTION_UNENROLL.send(
```
Comment (alex-sheehan-edx) [inline openedx_ledger/views.py:55]: I'm actually now wondering if we can leverage the reversal signal instead of needing the admin. Is there ever a point where a reversal shouldn't be followed by an un-enrollment? What if we just registered a receiver for `TRANSACTION_REVERSED` in subsidy that made the call to unenroll on platform?

File: openedx_ledger/api.py [hunk 12]
```diff
@@ -93,6 +94,7 @@ def reverse_full_transaction(transaction, idempotency_key, **metadata):
                 'metadata': metadata,
             },
         )
+        TRANSACTION_REVERSED.send(sender=None, reversal=reversal)
```
Comment (iloveagent57) [inline openedx_ledger/api.py:97]: nit: you probably want `sender=models.Reversal`

File: openedx_ledger/signals/signals.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+#     ]
+TRANSACTION_REVERSED = Signal()
+
+# Signal that indicates that an admin has invoked a transaction reversal/cancelation event.
+# providing_args=[
+#         'transaction',  # Transaction object
+#         'admin',  # User object
+#         'reversal',  # Optional Reversal object
+#     ]
+ADMIN_INVOKED_TRANSACTION_UNENROLL = Signal()
```
Comment (iloveagent57) [inline openedx_ledger/signals/signals.py:19]: You can delete this chunk.

File: openedx_ledger/views.py (hunk 11 above)
Comment (iloveagent57) [reply openedx_ledger/views.py:55]: right, you don't want `ADMIN_INVOKED_TRANSACTION_UNENROLL` to exist anymore.  But `reverse_full_transaction()` is going to send the `TRANSACTION_REVERSED` signal for you, you don't need to do it in this view.

**Review by iloveagent57 — COMMENTED** [review]
couple more things to clean up, but this is looking good!

File: openedx_ledger/views.py (hunk 6 above)
Comment (alex-sheehan-edx) [reply openedx_ledger/views.py:40]: this is no longer a concern with the current implementation

File: openedx_ledger/views.py [hunk 14]
```diff
@@ -1,6 +1,55 @@
...
-# Create your views here.
+from django.http import HttpResponseRedirect
+from django.shortcuts import render
+from django.urls import reverse
+from django.views.generic import View
+
+from openedx_ledger.api import reverse_full_transaction
+from openedx_ledger.models import Transaction
+
+logger = logging.getLogger(__name__)
```
Comment (iloveagent57) [inline openedx_ledger/views.py:14]: 🤦 thank you!

**Review by iloveagent57 — APPROVED** [review]
Approved!  You may want unit tests. ;)

**Review by johnnagro — APPROVED** [review]
⛵
